System/Random.hs

```diff
@@ -470,75 +486,144 @@ class RandomGen g where
           (h32, g'') ->
             ((fromIntegral h32 `unsafeShiftL` 32) .|. fromIntegral l32, g'')
 
+  -- | @genWord32R upperBound g@ returns a 'Word32' that is uniformly
+  -- distributed over the range @[0, upperBound]@.
+  --
+  -- @since 1.2
   genWord32R :: Word32 -> g -> (Word32, g)
   genWord32R m g = runGenState g (unbiasedWordMult32 m)
 
+  -- | @genWord64R upperBound g@ returns a 'Word64' that is uniformly
+  -- distributed over the range @[0, upperBound]@.
+  --
+  -- @since 1.2
   genWord64R :: Word64 -> g -> (Word64, g)
   genWord64R m g = runGenState g (unsignedBitmaskWithRejectionM uniformWord64 m)
 
+  -- | @genShortByteString n g@ returns a 'ShortByteString' of length @n@
+  -- filled with random bytes.
+  --
+  -- @since 1.2
   genShortByteString :: Int -> g -> (ShortByteString, g)
   genShortByteString n g =
     unsafePerformIO $ runGenStateT g (genShortByteStringWith n . uniformWord64)
   {-# INLINE genShortByteString #-}
-  -- |The 'genRange' operation yields the range of values returned by
-  -- the generator.
+
+  -- | Yields the range of values returned by 'next'.
   --
   -- It is required that:
   --
-  -- * If @(a, b) = 'genRange' g@, then @a <= b@.
-  --
-  -- * 'genRange' always returns a pair of defined 'Int's.
-  --
-  -- The second condition ensures that 'genRange' cannot examine its
-  -- argument, and hence the value it returns can be determined only by the
-  -- instance of 'RandomGen'.  That in turn allows an implementation to make
-  -- a single call to 'genRange' to establish a generator's range, without
-  -- being concerned that the generator returned by (say) 'next' might have
-  -- a different range to the generator passed to 'next'.
+  -- *   If @(a, b) = 'genRange' g@, then @a < b@.
+  -- *   'genRange' must not examine its argument so the value it returns is
+  --     determined only by the instance of 'RandomGen'.
   --
   -- The default definition spans the full range of 'Int'.
   genRange :: g -> (Int, Int)
   genRange _ = (minBound, maxBound)
 
-  -- | The 'split' operation allows one to obtain two distinct random number
-  -- generators.
+  -- | Returns two distinct pseudo-random number generators.
+  --
+  -- Implementations should take care to ensure that the resulting generators
+  -- are not correlated. Some pseudo-random number generators are not
+  -- splittable. In that case, the 'split' implementation should fail with a
+  -- descriptive 'error' message.
```

Comment:
Do you agree with this guidance? Is there a more elegant way to deal with this?

Reply:
I think the only elegant way to this problem is splitting RandomGen into two classes as discussed in #7 Until and if that will happen, the `error` is the best we can hope for.

Reply:
:+1: